File: 🟡 Medium/2326 - Spiral Matrix IV/_2326SpiralMatrixIV.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> res(m, vector<int>(n, -1));
        int tR = 0, bR = m - 1, lC = 0, rC = n - 1;
        // CW Direction -> top row --> right column --> bottom row --> left
        // column
        while (head != NULL) {
            int column = lC;
            while (column <= rC && head != NULL) {
                res[tR][column] = head->val;
                column++;
                head = head->next;
            }
            tR++;

            int row = tR;
            while (row <= bR && head != NULL) {
                res[row][rC] = head->val;
                row++;
                head = head->next;
            }
            rC--;

            column = rC;
            while (column >= lC && head != NULL) {
                res[bR][column] = head->val;
                column--;
                head = head->next;
            }
            bR--;

            row = bR;
            while (row >= tR && head != NULL) {
                res[row][lC] = head->val;
                row--;
                head = head->next;
            }
            lC++;
        }

        return res;
    }
};

File: 🟡 Medium/2326 - Spiral Matrix IV/_2326SpiralMatrixIV.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[][] spiralMatrix(int m, int n, ListNode head) {
        int res[][] = new int[m][n];
        for (int[] row : res) {
            Arrays.fill(row, -1);
        }

        int tR = 0, bR = m - 1, lC = 0, rC = n - 1;
        // CW Direction -> top row --> right column --> bottom row --> left column
        while (head != null) {
            int column = lC;
            while (column <= rC && head != null) {
                res[tR][column] = head.val;
                column++;
                head = head.next;
            }
            tR++;

            int row = tR;
            while (row <= bR && head != null) {
                res[row][rC] = head.val;
                row++;
                head = head.next;
            }
            rC--;

            column = rC;
            while (column >= lC && head != null) {
                res[bR][column] = head.val;
                column--;
                head = head.next;
            }
            bR--;

            row = bR;
            while (row >= tR && head != null) {
                res[row][lC] = head.val;
                row--;
                head = head.next;
            }
            lC++;
        }

        return res;
    }
}